Plasma roundup (#21)
kauri-io/Content: plasma-roundup/draft-plasma-article.md

## Prerequisites

For most of these projects you need Truffle, Ganache and npm installed.

## FourthState Labs

Among the early Plasma implementations was [FourthState Labs](https://github.com/FourthState/plasma-mvp-rootchain), whose design included a rootchain contract according to the Plasma MVP. This rootchain, which other projects have also incorporated, is a series of Solidity smart contracts. It may be a good place to start if you are considering building your own Plasma chain. It is designed to maintain a mapping from from block number to merkle root, processing deposits, withdrawals and resolving transaction disputes. FourthState has written tests for these features which you can run by following these steps:

```shell
git clone https://github.com/fourthstate/plasma-mvp-rootchain
cd plasma-mvp-rootchain
npm install
ganache-cli // In seperate terminal window
npm test
```

![FourthState tests](/images-for-article/Fourth-Estate/fourth-estate.png)

<!-- TODO: Running? -->

## OmiseGO

> Other notable early MVP's include [OmiseGO's research implementation](https://github.com/omisego/plasma-mvp), which has a root chain, child chain and a client to interact with the Plasma chain.

Reviewer: ```suggestion
Other notable early MVP's include [OmiseGO's research implementation](https://github.com/omisego/plasma-mvp), which has a root chain, child chain and a client to interact with the Plasma chain. Moving beyond research, [Kyokan](https://github.com/kyokan/plasma) introduced their Golang implementation [extending the original MVP specification](https://kauri.io/article/7f9e1c04f3964016806becc33003bdf3/v4/minimum-viable-plasma-the-kyokan-implementation). Kyokan uses the FourthState rootchain contract reviewed above. The architecture includes root nodes to process transactions and package them into blocks, broadcasts blocks to validator nodes, processes exits and more.
```

> Other notable early MVP's include [OmiseGO's research implementation](https://github.com/omisego/plasma-mvp), which has a root chain, child chain and a client to interact with the Plasma chain. Moving beyond research, [Kyokan](https://github.com/kyokan/plasma) introduced their Golang implementation [extending the original MVP specification](https://kauri.io/article/7f9e1c04f3964016806becc33003bdf3/v4/minimum-viable-plasma-the-kyokan-implementation). Kyokan uses the FourthState rootchain contract reviewed above. The architecture includes root nodes to process transactions and package them into blocks, broadcasts blocks to validator nodes, processes exits and more.

Reviewer: ```suggestion

```

> # Plasma Roundup: From MVP to Plasma Group

Reviewer: ```suggestion
# Plasma Roundup: From MVP to Mainnet
```

Reviewer: Think I need to update the title and overall angle a bit to reflect more recent Plasma progress.

In August 2017, Plasma creators Joseph Poon and Vitalik Buterin [proposed this framework](https://plasma.io/plasma.pdf) for scaling Ethereum transaction throughput to a "significant amount of state updates per second," potentially more than Paypal, Visa or other widely used merchant service providers.

Aside from offering comparable, if not more, transactions per second Plasma stakes its security in the value of Ethereum’s decentralized mainchain rather than in a centralized merchant service.

The promise of Plasma lays in its potential to help scale blockchain technology by processing a substantial amount of decentralized financial applications worldwide.

In January 2018, Vitalik posted the [“Minimal Viable Plasma” (MVP) specification to the Ethereum Research Forums](https://ethresear.ch/t/minimal-viable-plasma/426). The specification was designed to offer simplicity and basic security properties to kickstart development. Teams immediately began building their own implementations.

This overview is not meant to be exhaustive. Instead, by highlighting a few implementations, it aims to be indicative of the progress that Plasma has made over the past year.

> Although none of the Plasma models reviewed here are on mainnet or production ready, they show that this technology is not just theory. Taken together, the implementations suggest that Plasma is moving rapidly toward realizing the scaling potential that its creators and implementers envision.

Reviewer: ```suggestion
Although none of the Plasma models reviewed here are production ready, they show that this technology is not just theory. Taken together, the implementations suggest that Plasma is moving rapidly toward realizing the scaling potential that its creators and implementers envision.
```

## Prerequisites

For most of these projects you need Truffle, Ganache and npm installed.

## FourthState Labs

> Among the early Plasma implementations was [FourthState Labs](https://github.com/FourthState/plasma-mvp-rootchain), whose design included a rootchain contract according to the Plasma MVP. This rootchain, which other projects have also incorporated, is a series of Solidity smart contracts. Follow these steps to test them:

Reviewer: ```suggestion
Among the early Plasma implementations was [FourthState Labs](https://github.com/FourthState/plasma-mvp-rootchain), whose design included a rootchain contract according to the Plasma MVP. This rootchain, which other projects have also incorporated, is a series of Solidity smart contracts. It may be a good place to start if you are considering building your own Plasma chain. It is designed to maintain a mapping from from block number to merkle root, processing deposits, withdrawals and resolving transaction disputes. FourthState has written tests for these features which you can run by following these steps:
```

```shell
git clone https://github.com/fourthstate/plasma-mvp-rootchain
cd plasma-mvp-rootchain
npm install
ganache-cli // In seperate terminal window
npm test
```

![FourthState tests](/images-for-article/Fourth-Estate/fourth-estate.png)

<!-- TODO: Running? -->

## OmiseGO

Other notable early MVP's include [OmiseGO's implementation](https://github.com/omisego/plasma-contracts).  OmiseGO aspires to enable financial inclusion and interoperability through a public, decentralized OMG network. A key component of this network is Plasma. OMG’s implementation which has a root chain, child chain and a client to interact with the Plasma chain, is different from the MVP specification, however. For instance, OMG has added protection against the threat of chain re-organization (which can result from 51% attacks). And, among other additions, it has built in support for ERC20 token handling.

> Making the significant jump from MVP to mainnet is [Kyokan](https://github.com/kyokan/plasma). Kyokan is a Golang implementation [extending the original MVP specification](https://kauri.io/article/7f9e1c04f3964016806becc33003bdf3/v4/minimum-viable-plasma-the-kyokan-implementation). Their architecture uses the FourthState rootchain contract reviewed above and also includes root nodes to process transactions and package them into blocks, broadcast blocks to validator nodes, process exits and more. The team has been working steady since March 2018 to build a production-ready plasma implementation. Last month, just one year later, Kyokan announced that they have achieved [two critical milestones on the way to that goal](https://medium.com/kyokan-llc/announcing-our-plasma-mvp-alpha-23a8bc9673fc): the launch of their MVP mainnet alpha (capable of an initial 1,000 transactions per second) and the completion of their security audit. To help test how their Plasma will behave now that they have released it into the wild, Kyokan has launched a game of Capture the Flag where people are invited to break the system and keep the funds “just let us know how you did it, so that the rest of the community can benefit.” Successful denial of service attacks and attacks that lead to a loss of funds in [the smart contract](https://etherscan.io/address/0x0cdd78c34a4305234898864c1daccdbb326a520d) will also be paid $1,000.

Reviewer: ```suggestion
## Kyokan Plasma

Making the significant jump from MVP to mainnet is [Kyokan](https://github.com/kyokan/plasma). Kyokan is a Golang implementation [extending the original MVP specification](https://kauri.io/article/7f9e1c04f3964016806becc33003bdf3/v4/minimum-viable-plasma-the-kyokan-implementation). Their architecture uses the FourthState rootchain contract reviewed above and also includes root nodes to process transactions and package them into blocks, broadcast blocks to validator nodes, process exits and more. The team has been working steady since March 2018 to build a production-ready plasma implementation. Last month, just one year later, Kyokan announced that they have achieved [two critical milestones on the way to that goal](https://medium.com/kyokan-llc/announcing-our-plasma-mvp-alpha-23a8bc9673fc): the launch of their MVP mainnet alpha (capable of an initial 1,000 transactions per second) and the completion of their security audit. To help test how their Plasma will behave now that they have released it into the wild, Kyokan has launched a game of Capture the Flag where people are invited to break the system and keep the funds “just let us know how you did it, so that the rest of the community can benefit.” Successful denial of service attacks and attacks that lead to a loss of funds in [the smart contract](https://etherscan.io/address/0x0cdd78c34a4305234898864c1daccdbb326a520d) will also be paid $1,000.
```

While Plasma’s arrival on mainnnet is a notable milestone, the technology’s evolution needs more work and time before it is ready for mass adoption.  

> ## Kyokan Plasma

Reviewer: ```suggestion

```

While Plasma’s arrival on mainnnet is a notable milestone, the technology’s evolution needs more work and time before it is ready for mass adoption.  

> Moving beyond research, [Kyokan](https://github.com/kyokan/plasma) introduced their Golang implementation [extending the original MVP specification](https://kauri.io/article/7f9e1c04f3964016806becc33003bdf3/v4/minimum-viable-plasma-the-kyokan-implementation). Kyokan uses the FourthState rootchain contract reviewed above. The architecture includes root nodes to process transactions and package them into blocks, broadcasts blocks to validator nodes, processes exits and more.

Reviewer: ```suggestion

```

```shell
git clone https://github.com/fourthstate/plasma-mvp-rootchain
cd plasma-mvp-rootchain
npm install
ganache-cli // In seperate terminal window
npm test
```

![FourthState tests](/images-for-article/Fourth-Estate/fourth-estate.png)

<!-- TODO: Running? -->

## OmiseGO

Other notable early MVP's include [OmiseGO's implementation](https://github.com/omisego/plasma-contracts).  OmiseGO aspires to enable financial inclusion and interoperability through a public, decentralized OMG network. A key component of this network is Plasma. OMG’s implementation which has a root chain, child chain and a client to interact with the Plasma chain, is different from the MVP specification, however. For instance, OMG has added protection against the threat of chain re-organization (which can result from 51% attacks). And, among other additions, it has built in support for ERC20 token handling.

## Kyokan Plasma

> Making the significant jump from MVP to mainnet is [Kyokan](https://github.com/kyokan/plasma). Kyokan is a Golang implementation [extending the original MVP specification](https://kauri.io/article/7f9e1c04f3964016806becc33003bdf3/v4/minimum-viable-plasma-the-kyokan-implementation). Their architecture uses the FourthState rootchain contract reviewed above and also includes root nodes to process transactions and package them into blocks, broadcast blocks to validator nodes, process exits and more. The team has been working steady since March 2018 to build a production-ready plasma implementation. Last month, just one year later, Kyokan announced that they have achieved [two critical milestones on the way to that goal](https://medium.com/kyokan-llc/announcing-our-plasma-mvp-alpha-23a8bc9673fc): the launch of their MVP mainnet alpha (capable of an initial 1,000 transactions per second) and the completion of their security audit. To help test how their Plasma will behave now that they have released it into the wild, Kyokan has launched a game of Capture the Flag where people are invited to break the system and keep the funds “just let us know how you did it, so that the rest of the community can benefit.” Successful denial of service attacks and attacks that lead to a loss of funds in [the smart contract](https://etherscan.io/address/0x0cdd78c34a4305234898864c1daccdbb326a520d) will also be paid $1,000.

Reviewer: ```suggestion
Making the significant jump from MVP to mainnet is [Kyokan](https://github.com/kyokan/plasma). Kyokan is a Golang implementation [extending the original MVP specification](https://kauri.io/article/7f9e1c04f3964016806becc33003bdf3/v4/minimum-viable-plasma-the-kyokan-implementation). Their architecture uses the FourthState rootchain contract reviewed above and also includes root nodes to process transactions and package them into blocks, broadcast blocks to validator nodes, process exits and more. The team has been working steady since March 2018 to build a production-ready plasma implementation. Last month, Kyokan announced that they have achieved [two critical milestones on the way to that goal](https://medium.com/kyokan-llc/announcing-our-plasma-mvp-alpha-23a8bc9673fc): the launch of their MVP mainnet alpha (capable of an initial 1,000 transactions per second) and the completion of their security audit. To help test how their implementation will behave now that they have released it into the wild, Kyokan has launched a game of Capture the Flag where people are invited to break the system and keep the funds, “just let us know how you did it, so that the rest of the community can benefit.” Successful denial of service attacks and attacks that lead to a loss of funds in [the smart contract](https://etherscan.io/address/0x0cdd78c34a4305234898864c1daccdbb326a520d) will also be paid $1,000.
```

In August 2017, Plasma creators Joseph Poon and Vitalik Buterin [proposed this framework](https://plasma.io/plasma.pdf) for scaling Ethereum transaction throughput to a "significant amount of state updates per second," potentially more than Paypal, Visa or other widely used merchant service providers.

Aside from offering comparable, if not more, transactions per second Plasma stakes its security in the value of Ethereum’s decentralized mainchain rather than in a centralized merchant service.

The promise of Plasma lays in its potential to help scale blockchain technology by processing a substantial amount of decentralized financial applications worldwide.

In January 2018, Vitalik posted the [“Minimal Viable Plasma” (MVP) specification to the Ethereum Research Forums](https://ethresear.ch/t/minimal-viable-plasma/426). The specification was designed to offer simplicity and basic security properties to kickstart development. Teams immediately began building their own implementations.

This overview is not meant to be exhaustive. Instead, by highlighting a few implementations, it aims to be indicative of the progress that Plasma has made over the past year.

> Although none of the Plasma models reviewed here are on mainnet or production ready, they demonstrate that this technology is not just theory. Taken together, the many implementations suggest that Plasma is moving rapidly toward realizing the scaling potential that its creators and implementers envision.     

Reviewer: ```suggestion
Although none of the Plasma models reviewed here are production ready, they demonstrate that this technology is not just theory. Taken together, the many implementations suggest that Plasma is moving rapidly toward realizing the scaling potential that its creators and implementers envision.     
```

> Among the early Plasma implementations was [FourthState Labs](https://github.com/FourthState/plasma-mvp-rootchain), whose design included a rootchain contract according to the Plasma MVP. This rootchain, which has been incorporated into other implementations, is pretty simple to test. To test, I used this environment:

Reviewer: ```suggestion
Among the early Plasma implementations was [FourthState Labs](https://github.com/FourthState/plasma-mvp-rootchain), whose design included a rootchain contract according to the Plasma MVP. This rootchain, which other projects have also incorporated, is a series of Solidity smart contracts. It may be a good place to start if you are considering building your own Plasma chain. As a rootchain it is designed to maintain a mapping from block number to merkle root, processing deposits, withdrawals and resolving transaction disputes. FourthState has written tests for these features. I tested them in this environment:
To test, I used this environment:
```

* Linux (Debian Stretch)
* Truffle (v5.0.2)
* Ganache CLI (v6.2.5)
* NPM 6.7.0

> Follow these steps to test:

Author: @cspannos A lot of my comments might be more related to the state of these projects, than your work…

But why would I want to test? And looking into their code, I'm not even sure the test does anything constructive. How can I actually run things? Is that possible? I know their docs are lacking, but maybe we can fill that gap if you have any ideas at all?

Reviewer: @ChrisChinchilla Great questions! I will take another look at their repo to see if anything can be added there. For now, my latest draft attempted to answer only at a very high level. If you think it is insufficient, however, I am absolutely fine if my reference to testing is deleted.

0. `$ git clone https://github.com/fourthstate/plasma-mvp-rootchain`
1. `$ cd plasma-mvp-rootchain`
2. `$ npm install`
3. `$ npm install -g truffle ganache-cli` // if not installed already
4. `$ ganache-cli` // run as a background process
5. `$ npm test`

![FourthState tests](/images-for-article/Fourth-Estate/fourth-estate.png)

> Other notable early MVP's include [OmiseGO's research implementation](https://github.com/omisego/plasma-mvp), which has a root chain, child chain and a client to interact with the Plasma chain. Moving beyond research, [Kyokan](https://github.com/kyokan/plasma) introduced their Golang implementation [extending the original MVP specification](https://kauri.io/article/7f9e1c04f3964016806becc33003bdf3/v4/minimum-viable-plasma-the-kyokan-implementation). Kyokan uses the FourthState rootchain contract reviewed above. The architecture includes root nodes to process transactions and package them into blocks, broadcasts blocks to validator nodes, processes exits and more.

Author: @cspannos OmiseGo seems to have moved here - https://github.com/omisego/plasma-contracts

Any any chance of a bit more detail?

Reviewer: @ChrisChinchilla Yes, absolutely. I took a crack at it. Let me know if you think there should be more or if you have anything specific in mind.

> Other notable early MVP's include [OmiseGO's research implementation](https://github.com/omisego/plasma-mvp), which has a root chain, child chain and a client to interact with the Plasma chain. Moving beyond research, [Kyokan](https://github.com/kyokan/plasma) introduced their Golang implementation [extending the original MVP specification](https://kauri.io/article/7f9e1c04f3964016806becc33003bdf3/v4/minimum-viable-plasma-the-kyokan-implementation). Kyokan uses the FourthState rootchain contract reviewed above. The architecture includes root nodes to process transactions and package them into blocks, broadcasts blocks to validator nodes, processes exits and more.

Reviewer: ```suggestion
Other notable early MVP's include [OmiseGO's implementation](https://github.com/omisego/plasma-contracts).  OmiseGO aspires to enable financial inclusion and interoperability through a public, decentralized OMG network. A key component of this network is Plasma. OMG’s implementation which has a root chain, child chain and a client to interact with the Plasma chain, is different from the MVP specification, however. For instance, OMG has added protection against the threat of chain re-organization (which can result from 51% attacks). And, among other additions, it has built in support for ERC20 token handling.

Making the significant jump from MVP to mainnet is [Kyokan](https://github.com/kyokan/plasma). Kyokan is a Golang implementation [extending the original MVP specification](https://kauri.io/article/7f9e1c04f3964016806becc33003bdf3/v4/minimum-viable-plasma-the-kyokan-implementation). Their architecture uses the FourthState rootchain contract reviewed above and also includes root nodes to process transactions and package them into blocks, broadcast blocks to validator nodes, process exits and more. The team has been working steady since March 2018 to build a production-ready plasma implementation. Last month, just one year later, Kyokan announced that they have achieved [two critical milestones on the way to that goal](https://medium.com/kyokan-llc/announcing-our-plasma-mvp-alpha-23a8bc9673fc): the launch of their MVP mainnet alpha (capable of an initial 1,000 transactions per second) and the completion of their security audit. To help test how their Plasma will behave now that they have released it into the wild, Kyokan has launched a game of Capture the Flag where people are invited to break the system and keep the funds “just let us know how you did it, so that the rest of the community can benefit.” Successful denial of service attacks and attacks that lead to a loss of funds in [the smart contract](https://etherscan.io/address/0x0cdd78c34a4305234898864c1daccdbb326a520d) will also be paid $1,000.

While Plasma’s arrival on mainnnet is a notable milestone, the technology’s evolution needs more work and time before it is ready for mass adoption.  

```

> In Kyokan, validator nodes check the validity of blocks and exit if bad behavior is detected. The Plasma contract lives on the Ethereum root chain and supports deposits, block submissions, exits and challenges. Kyokan, which aims to provide a pluggable architecture, has deployed their [Plasma Block Explorer](https://explorer.kyokan.io/) on the Rinkeby test net:

Reviewer: ```suggestion

```

Reviewer: Can do without this para.

> ![Kyokan Block Explorer](/images-for-article/Kyokan/kyokan-block-explorer.png)

Reviewer: ```suggestion
```

Reviewer: Screenshot of block explorer no longer useful.

> Closing 2018, the [Plasma Group](https://plasma.group/) [announced the release of their implementation](https://medium.com/plasma-group/plasma-spec-9d98d0f2fccf) aimed at the greater Ethereum community. It includes a Plasma chain operator, a client and command line wallet, support for ERC20 tokens, a block explorer, transaction load testing and more.

Reviewer: ```suggestion
Closing 2018, the [Plasma Group](https://plasma.group/) [announced the release of their implementation](https://medium.com/plasma-group/plasma-spec-9d98d0f2fccf) aimed at the greater Ethereum community. It includes a Plasma chain operator, a client and command line wallet, support for ERC20 tokens, a block explorer, transaction load testing and more. While their implementation includes properties such as scalable light client proofs and the possibility for interchain atomic swaps, the group has moved quickly to offer a [general purpose plasma design](https://medium.com/plasma-group/towards-a-general-purpose-plasma-f1cc4d49c1f4). This general purpose design aims to overcome constraints in old design patterns which were not upgradeable and not generalizable. The purpose is to create a plasma environment that would allow the development of applications on a Plasma chain in the same way that dapps are built on the Ethereum blockchain. Hence the Plasma Group has recently offered plasma apps, also known as ‘plapps’. Plapps are a special type of smart contract called a ‘[predicate contract](https://github.com/plasma-group/plasma-predicates)’. Plapps are deployed to Ethereum, where users can interact with the contract and the plasma chain carries out the computation, which is less expensive than if it were to occur on the Ethereum main chain. 
```

> Many aspects of [this implementation are currently testable](https://github.com/plasma-group). To test the [Plasma Core](https://github.com/plasma-group/plasma-core) (again using the Debian, Ganache, NPM environment), follow these steps:

Author: @cspannos Again, why do I want to run the project tests? As far as I can tell, this is running code tests, not me "testing" the project, i.e. "running it".

Reviewer: I think I was excited by what they were offering at the time :) I removed my reference to testing in the latest version. However, I did keep the stuff on the block explorer. Again, I am happy to delete it though, if it seems superfluous!

> Many aspects of [this implementation are currently testable](https://github.com/plasma-group). To test the [Plasma Core](https://github.com/plasma-group/plasma-core) (again using the Debian, Ganache, NPM environment), follow these steps:

Reviewer: ```suggestion
```
